### ISSUE TITLE
Background File Syncing

### ISSUE THREAD

**Issue by axelu** [strongbox-password-safe/Strongbox#81]
The current iOS implementation only downloads a safe from the remote on opening the safe.  

The new desired behavior is, that it would automatically every so often, frequency ideally configurable with a sensible default, refresh from the remote. This assumes the app is running, like running in the background.  That setting itself should be an option to be enabled or not.  

On the list of safes it should show the date/time stamp of the local safe so that one knows, from what date/time the local copy is. Implementation wise maybe think of how Apple Mail on iOS shows last updated date/time when in a mailbox list screen like Inbox, but in this case specific to the safe. 

So, basically two enhancements:  
- show date/time of local safe on safe list screen
- new feature to enable periodic background download from remote with the frequency being configurable.  

Conflict resolution strategy to be determined.

**Comment by mmcguill** [comment]
Hi @axelu - Just a thought on this. I'm not sure where you're storing your database, but if you add your database to Strongbox via iOS Files rather than one of the native providers you will probably get improved synchronization. Apple manages this process in the background and from reports I've heard it is reasonably good/seamless... I've also heard reports the sync is not perfect or can be delayed... YMMV. Worth trying out?

**Comment by axelu** [comment]
@mmcguill - I am storing the database on a local SFTP server on my home network (see also #80).  Based on your suggestion I did once again search for a solution outside of Strongbox to get the file local and synced, but I am still coming up empty.  Closest maybe [Readdle Documents](https://readdle.com/documents), but even then it states [background sync is not supported](https://helpspot.readdle.com/documents/index.php?pg=kb.page&id=1203), so now one would rely on that app to be opened to trigger sync.  While going down the third party route to manage the sync of a database that is local is certainly possible, it feels clunky, even though it may bring 2-way sync.  
I still think that offering the ability to enable periodic background download from remote as a feature is elegant (it would be 1-way sync) and it would make Strongbox more complete and it would make Strongbox stand by itself without relying on some other mechanism to manage the databases.  Maybe the way to position it is as built-in 1-way sync, meaning periodic background download from remote when said remote is available.  It still leaves conflict resolution strategy to be determined.

**Comment by georgesnow** [comment]
How are hosting your sftp server? Are you using NAS? (Synology or qnap)

**Comment by shad0whawk** [comment]
Two-way entry-level compare and sync would be nice, but it doesn’t exist at the moment. Downloading the safe periodically in the background from the remote location to your device seems pointless if you’re not actively using it. When you do need to use it, why not just open and refresh it at that time? This requires one download versus multiple over time to achieve the same result. What benefit do you gain by implementing a periodic remote fetch process without any syncing mechanism? Just curious...

FWIW, you could host your database on a shared cloud service and link to it directly within Strongbox. This will sync your local changes with the cloud database and vice versa. Of course, that means it’s no longer hosted only on your own network. Your other devices can also open and modify the same database.

**Comment by axelu** [comment]
> @shad0whawk - Downloading the safe periodically in the background from the remote location to your device seems pointless if you’re not actively using it. When you do need to use it, why not just open and refresh it at that time? 

The point actually is, that there is no guarantee that the remote is accessible the moment one wants to use the database, meaning, it could not be refreshed at the time of opening.

> .@shad0whawk - .. you could host your database on a shared cloud service ...

I do not want to host the database on a shared cloud service.  The database is really only manipulated on one device which then pushes the changed database up to the NAS, and then the database shall be available on a few other devices including my iPhone (as read only).  I have it downloaded on at least one other device periodically via simple cron script.

> @georgesnow - How are hosting your sftp server? Are you using NAS? (Synology or qnap)

Its a home-built NAS not using a pre-configured package.  Why are you asking?  If there is a suggestion to use something, I can install/configure/script anything on my NAS.

**Comment by georgesnow** [comment]
Well I have Synology NAS and I have tested strongbox  using cloud station sync(ds cloud) with a database. It works basicly like your own Dropbox. Also My Synology isn’t accessible outside my house. It only syncs when I am out home.

However your case is a bit different. Wonder if you could use the iOS short cuts. You can run a script over ssh presumably you have ssh on if you have sftp and then have it open strongbox for you. 

The shortcut would You download a copy of the database to your device. And you might be able to specific /iPhone/shortcuts cause I think default it’s /iCloud/shortcuts. this would Make it accessible from Files app. I have to think about it more cause strongbox might view it as “new “ dB each time you download it. 

I do have a few other remote file apps I will see if any these provide a sync option

**Comment by mmcguill** [comment]
Hi @axelu are you aware of the Offline Cache feature? 

You can Slide Left on your database and tap 'Open Offline'? 

Does that ameliorate things?

The offline cache gets updated any time you unlock/save while online. So it will have the latest known version that way. Many people find this useful.

On background refresh, this is not really feasible for iOS apps, they basically get suspended/hibernated after about 2-3mins, so there's no real way to keep watching something or regularly polling. This fact even makes clearing the Clipboard feature difficult.

Re: The Date/Time on main databases list screen I've created #130 to track

**Comment by axelu** [comment]
@mmcguill 

> are you aware of the Offline Cache feature?

I am aware of the Offline Cache feature and I am using it.  It works great and as desired.  Also, even if I have not enabled the Open Offline option, if the remote is not accessible, it prompts me accordingly to confirm that I want to open the local cached version.  This is perfect.

My issue remains however, that the offline cache gets updated ***only*** when unlocking/saving under the condition that the remote is available at that very moment. 

> On background refresh, this is not really feasible for iOS apps, they basically get suspended/hibernated

There is the option for the app to register for Background App Refresh.  With that it could periodically try to fetch from remote.  This is what I suggest it should do.
[Background App Refresh - General](https://support.apple.com/en-us/HT202070)
[Background App Refresh - Developer](https://developer.apple.com/documentation/uikit/app_and_scenes/preparing_your_ui_to_run_in_the_background/updating_your_app_with_background_app_refresh)

**Comment by mmcguill** [comment]
Hi @axelu - While this is technically possible, I'm not really sure it's one of the App's core functionalities, or should be. This seems like a job for some other kind of background sync tool. 

Is your database very large in size or something?

Can you use iOS Files to access it? iOS Files manages sync differently (perhaps better in some ways/cases) than Strongbox native sync.

**Comment by georgesnow** [comment]
Hopefully not to confuse things, but my experience with sftp is local caching allows for editing while connection is “alive” and is only saved when user does so. example config (text) file on a device/server a local cache allows the text editing to modify and make changes but leaving the server file “untouched” until the user actually saves the file which writes it back to the remote at that moment. There are apps for macos like mountain duck that provide a sync like experience over sftp. But iOS I don’t know know of one

**Comment by georgesnow** [comment]
Sorry I didn't finish my thought. The behavior your describing in general is the expected behavior for sftp.

**Comment by axelu** [comment]
@mmcguill 

> Is your database very large in size or something?

No.

> Can you use iOS Files to access it? iOS Files manages sync differently (perhaps better in some ways/cases) than Strongbox native sync.

iOS Files does not natively support SFTP locations.  One has to look for an application that supports SFTP and integrates with iOS Files such as [FileBrowser](
https://www.stratospherix.com/products/filebrowser/).  FileBrowser has a sync feature that will periodically pull the file from the remote to Local on the iPhone and then in Strongbox one could open the synced file from Local.  I tried that and it works.  This would in essence solve my use case.  

While I don't disagree with you saying that a periodic sync in background is not core functionality of the app, making this the job of some other background sync tool seems, for me as a minimalist, very kludgy.  One has to get familiar with Files, find something that supports SFTP integrates with Files, manage that other app now, etc.  I understand and just now found a solution in FileBrowser that would work, I still think it would be a fine addition to the app to support background sync, maybe at least for SFTP, which is not natively in Files otherwise.  Obviously your call.

**Comment by mmcguill** [comment]
Thanks @axelu - I'm glad you found a solution. I do think there are improvements to be made in Caching for sure, I'm just not keen about having Strongbox (A Password Manager) do background file syncing. File syncing makes sense when the App is active perhaps, but the background aspect seems out of scope to me. It would also involve significant technical changes, so while not completely excluding the possibility of this, it definitely for me, sits down the priority list.

Hoping you can get a good solution working outside of this.

**Comment by mmcguill** [comment]
Closing this as Won't Fix - I feel this is out of scope
